### PR TITLE
feat: Switch TrieUpdate API to use TrieKey trait

### LINKED ISSUE #2276: Cleanup RuntimeExt iterator methods
The iterators were removed from `VMLogic`, but they also needs to be removed from `RuntimeExt`.

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
Introducing `TrieKey` trait to work with keys within `TrieUpdate` instead of using raw `Vec<u8>`.

This allows to have additional logic for keys. E.g. if we want to hash AccountID for keys (see https://github.com/nearprotocol/nearcore/issues/2215) and later return all touched accounts for a block (see https://github.com/nearprotocol/nearcore/pull/2253) we need a way to map `sha256(account_id)` from a raw key back to the `account_id`.

Then `TrieKey` may implement `fn get_account_id_with_hash() -> Option<(AccountId, CryptoHash)>` logic that provides Account ID and hash of this Account ID. With this info we can remember pairs of touched Accounts in memory of `TrieUpdate` and now we can map back hashes here: https://github.com/nearprotocol/nearcore/blob/62f00b211915790d06e822da866fb01ceb6cdc89/core/store/src/trie/mod.rs#L578

Removes iterators from VMLogic (cause they somewhat block changes from this PR).
Fixes #2276 

##Test Plan:
- unit tests and CI

### CODE REVIEW

File: core/primitives/src/utils.rs [hunk 1]
```diff
@@ -83,10 +105,11 @@ impl KeyForColumnAccountId {
     }
 }
 
-#[derive(derive_more::AsRef, derive_more::Into)]
-#[as_ref(forward)]
+#[derive(derive_more::Into)]
```
Comment (evgenykuzyakov) [inline core/primitives/src/utils.rs]: I guess I don't need `into_vec()` or `derive_more::Into`

File: runtime/near-vm-logic/src/dependencies.rs [hunk 2]
```diff
@@ -142,6 +142,7 @@ pub trait External {
     /// ```
     fn storage_has_key(&mut self, key: &[u8]) -> Result<bool>;
 
+    /*
```
Comment (frol) [inline runtime/near-vm-logic/src/dependencies.rs]: Do you plan to remove this code that you commented out?

File: runtime/near-vm-logic/src/logic.rs [hunk 3]
```diff
@@ -51,11 +50,6 @@ pub struct VMLogic<'a> {
     /// host-guest boundary.
     registers: HashMap<u64, Vec<u8>>,
 
-    /// Iterators that were created and can still be used.
```
Comment (frol) [inline runtime/near-vm-logic/src/logic.rs]: How is it related to the PR? Just a cleanup? I am not familiar with the module, so I don't know if it is ok to remove this code, but it seems that it was not used after all.

File: runtime/runtime/src/balance_checker.rs [hunk 4]
```diff
@@ -31,9 +32,9 @@ pub(crate) fn check_balance(
 ) -> Result<(), RuntimeError> {
     // Delayed receipts
     let initial_delayed_receipt_indices: DelayedReceiptIndices =
-        get(&initial_state, DELAYED_RECEIPT_INDICES)?.unwrap_or_default();
+        get(&initial_state, &KeyForDelayedReceiptIndices::new())?.unwrap_or_default();
```
Comment (frol) [inline runtime/runtime/src/balance_checker.rs]: :+1:

File: runtime/runtime/src/ext.rs [hunk 5]
```diff
@@ -65,10 +58,8 @@ impl<'a> RuntimeExt<'a> {
         }
     }
 
-    pub fn create_storage_key(&self, key: &[u8]) -> Vec<u8> {
-        let mut storage_key = self.storage_prefix.clone();
-        storage_key.extend_from_slice(key);
-        storage_key
+    pub fn create_storage_key(&self, key: &[u8]) -> KeyForData {
+        self.storage_prefix.with_suffix(&key)
```
Comment (frol) [inline runtime/runtime/src/ext.rs]: Type safety ftw!

File: core/primitives/src/utils.rs [hunk 6]
```diff
@@ -114,15 +137,21 @@ impl KeyForAccount {
...
+impl_trie_key!(KeyForAccessKey);
+
 impl KeyForAccessKey {
     fn estimate_prefix_len(account_id: &AccountId) -> usize {
         KeyForColumnAccountId::estimate_len(col::ACCESS_KEY, account_id) + col::ACCESS_KEY.len()
     }
 
+    /// This is not safe and should only be used internally for iterating over access keys for reading.
+    pub fn from_raw_key(key: &[u8]) -> Self {
+        Self(key.to_vec())
```
Comment (frol) [inline core/primitives/src/utils.rs]: I believe we can avoid this if we refactor the only place where this is needed: `NightShadeRuntime::view_access_keys`. Given that we create a `KeyForAccessKey` with `get_prefix` there, filtering by such a key should yield records with a key that is of `KeyForAccessKey` type.

Strictly speaking, this method should be marked as `unsafe` from the API point of view as it requires some external contract to be upheld (just like any `*_unchecked` method in Rust std library). I understand your hesitation to introduce unsafe code, I have been there just a few days ago, and ended up sourcing the iterator handling into the `KeyForStateChanges`:

https://github.com/nearprotocol/nearcore/blob/62f00b211915790d06e822da866fb01ceb6cdc89/core/store/src/trie/mod.rs#L622-L639

I am not 100% happy with it as I had to pollute the `KeyFor*` implementation with something a bit orthogonal, but I did not need to "hide" or introduce unsafe `from_raw_key` method.

File: core/primitives/src/utils.rs [hunk 7]
```diff
@@ -213,13 +243,25 @@ impl KeyForData {
         Self::get_prefix_with_capacity(account_id, 0)
     }
 
+    pub fn with_suffix(&self, suffix: &[u8]) -> Self {
+        let mut key = self.clone();
+        key.0.extend(suffix);
```
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply core/primitives/src/utils.rs]: Clone will be preferred in the future, once you have to store unhashed `AccountId` in it, but I can add this later

File: core/primitives/src/utils.rs [hunk 8]
```diff
@@ -52,9 +52,31 @@ pub mod col {
...
+            }
+        }
+
+        impl TrieKey for $iden {
+            fn into_vec(self) -> Vec<u8> {
+                self.0
+            }
+        }
+    };
+}
```
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply core/primitives/src/utils.rs]: Makes sense

File: runtime/near-vm-logic/src/dependencies.rs (hunk 2 above)
Comment (evgenykuzyakov) [reply runtime/near-vm-logic/src/dependencies.rs]: Yep. Might be leftover

File: runtime/near-vm-logic/src/logic.rs (hunk 3 above)
Comment (evgenykuzyakov) [reply runtime/near-vm-logic/src/logic.rs]: We've removed iterators from the VMLogic, but didn't cleanup the code. It affects some functions from `TrieUpdate`. So removing iterator code helps.

File: core/primitives/src/utils.rs (hunk 6 above)
Comment (evgenykuzyakov) [reply core/primitives/src/utils.rs]: I think it's okay to have it here until we remove iterators completely. Once we remove iterators and replace them with appropriate implementations based on unordered key-value store, we can remove `from_raw_key` methods.

File: runtime/near-vm-logic/src/logic.rs (hunk 3 above)
Comment (evgenykuzyakov) [reply runtime/near-vm-logic/src/logic.rs]: Fixes #2276

**Comment by nearmax** [comment]
This broke near-sdk-rs. See https://travis-ci.com/github/near/near-sdk-rs/jobs/305089218 I am fixing it

**Comment by nearmax** [comment]
We also did not build `0.6.1` with the correct code. See https://docs.rs/near-vm-logic/0.6.1/near_vm_logic/mocks/mock_external/struct.MockedExternal.html It corresponds to the code before the PR.

**Pull request merged**